tls/s2n_ktls.c

```diff
@@ -0,0 +1,199 @@
+{
+    RESULT_ENSURE_REF(conn);
+    RESULT_ENSURE_REF(conn->config);
+    RESULT_ENSURE_REF(conn->secure);
+    RESULT_ENSURE_REF(conn->secure->cipher_suite);
+    RESULT_ENSURE_REF(conn->secure->cipher_suite->record_alg);
+    RESULT_ENSURE_REF(conn->secure->cipher_suite->record_alg->cipher);
+
+    /* TODO support TLS 1.3
+     *
+     * TLS 1.3 support requires sending the KeyUpdate message when the cryptographic
+     * KeyLimits are met. However, this is currently only possible by applying a
+     * kernel patch to support this functionality.
+     */
+    RESULT_ENSURE_EQ(conn->actual_protocol_version, S2N_TLS12);
+
+    /* Check if the cipher supports kTLS */
+    RESULT_ENSURE_EQ(conn->secure->cipher_suite->record_alg->cipher->ktls_supported, true);
+    /* TODO key length is a weak check. Is there a better mechanism? */
+    uint8_t key_size = conn->secure->cipher_suite->record_alg->cipher->key_material_size;
+    /* only AES_GCM_128 is supported at the moment */
+    RESULT_ENSURE_EQ(key_size, S2N_TLS_AES_128_GCM_KEY_LEN);
+
+    return S2N_RESULT_OK;
+}
+
+S2N_RESULT s2n_ktls_validate_socket_mode(struct s2n_connection *conn, s2n_ktls_mode ktls_mode)
+{
+    RESULT_ENSURE_REF(conn);
+    RESULT_ENSURE(ktls_mode == S2N_KTLS_MODE_RECV || ktls_mode == S2N_KTLS_MODE_SEND, S2N_ERR_SAFETY);
+
+    /* kTLS I/O functionality is managed by s2n-tls. kTLS cannot be enabled if the
+     * application sets custom I/O (managed_send_io == false means application has
+     * set custom I/O).
+     *
+     * - Safety checks
+     * - Confirm application is not using custom I/O
+     * - Confirm kTLS isn't enabled already
+     */
+    switch (ktls_mode) {
+        case S2N_KTLS_MODE_SEND:
+            RESULT_ENSURE_REF(conn->send_io_context);
+            RESULT_ENSURE(!conn->managed_send_io, S2N_ERR_KTLS);
+            RESULT_ENSURE(!conn->ktls_send_enabled, S2N_ERR_KTLS);
+            break;
+        case S2N_KTLS_MODE_RECV:
+            RESULT_ENSURE_REF(conn->recv_io_context);
+            RESULT_ENSURE(!conn->managed_recv_io, S2N_ERR_KTLS);
+            RESULT_ENSURE(!conn->ktls_recv_enabled, S2N_ERR_KTLS);
+            break;
+        case S2N_KTLS_MODE_DISABLED:
+        case S2N_KTLS_MODE_DUPLEX:
+            RESULT_BAIL(S2N_ERR_SAFETY);
+            break;
+    }
+
+    return S2N_RESULT_OK;
+}
+
+S2N_RESULT s2n_ktls_retrieve_file_descriptor(struct s2n_connection *conn, s2n_ktls_mode ktls_mode, int *fd)
+{
+    RESULT_GUARD(s2n_ktls_validate_socket_mode(conn, ktls_mode));
+    RESULT_ENSURE_REF(fd);
+
+    if (ktls_mode == S2N_KTLS_MODE_RECV) {
+        /* retrieve the receive fd */
+        const struct s2n_socket_read_io_context *peer_socket_ctx = conn->recv_io_context;
+        *fd = peer_socket_ctx->fd;
+    } else if (ktls_mode == S2N_KTLS_MODE_SEND) {
+        /* retrieve the send fd */
+        const struct s2n_socket_write_io_context *peer_socket_ctx = conn->send_io_context;
+        *fd = peer_socket_ctx->fd;
```

Comment:
It's dangerous to rely on another method (s2n_ktls_validate_socket_mode) to do your null checking for you. The implementation of that method can change without noticing the assumptions that you're making in this method.

But you're also assuming the type of send_io_context because of s2n_ktls_validate_socket_mode. I would recombine the two methods so that your safety checks are near the code that assumes that they're true.

```suggestion
        /* retrieve the send fd */
        const struct s2n_socket_write_io_context *peer_socket_ctx = conn->send_io_context;
        RESULT_ENSURE_REF(peer_socket_ctx);
        *fd = peer_socket_ctx->fd;
```